### ISSUE TITLE
Cannot compile on Debian due to GLFW error

### ISSUE THREAD

**Issue by chandler-latour** [xtreme8000/aof-opengl#108]
Hello. I've had this error happen to me on two different Debian machines when trying to compile this thing, so I decided to log it as an issue, even though it might be something I'm doing wrong.
I always get this error message that doesn't seem to have anything to do with me missing a library.

Here's the error message:  
`
/home/blehmeh98/games/aceofspades/betterspades/BetterSpades/src/window.c:177:5:warning: implicit declaration of function ‘glfwRawMouseMotionSupported’; did you mean ‘glfwExtensionSupported’? [-Wimplicit-function-declaration]
  if(glfwRawMouseMotionSupported())
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~
     glfwExtensionSupported
/home/blehmeh98/games/aceofspades/betterspades/BetterSpades/src/window.c:178:38error: ‘GLFW_RAW_MOUSE_MOTION’ undeclared (first use in this function); did you mean ‘GLFW_MOUSE_BUTTON_4’?
   glfwSetInputMode(hud_window->impl, GLFW_RAW_MOUSE_MOTION, GLFW_TRUE);
                                      ^~~~~~~~~~~~~~~~~~~~~
                                      GLFW_MOUSE_BUTTON_4
/home/blehmeh98/games/aceofspades/betterspades/BetterSpades/src/window.c:178:38note: each undeclared identifier is reported only once for each function it appears in
make[2]: *** [src/CMakeFiles/client.dir/build.make:414: src/CMakeFiles/client.dir/window.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:91: src/CMakeFiles/client.dir/all] Error 2
`

I've tried compiling this on my Debian 10 laptop and on my Raspberry Pi running Raspian 10 Lite, and on both machines, I get this same exact error despite double checking that I have all the libraries installed. It's always "GLFW_RAW_MOUSE_MOTION undeclared", and it always suggests some mouse button instead.
I have libglfw3-dev installed. I tried installing libglfw-dev instead, but then it complains that it can't find GLFW3.h or something.

Is there an easy fix for this? Did I do something wrong? Or did GLFW_RAW_MOUSE_MOTION get depreciated at some point and now we're all screwed until we find a replacement?

**Comment by xtreme8000** [comment]
Debian's libglfw3 packages are very outdated (I just looked, in fact their version suggests they're from 2016).
I always build from their latest master: https://github.com/glfw/glfw

You need atleast version 3.3 while debians is 3.2.1

**Comment by chandler-latour** [comment]
I git cloned glfw from source, compiled and installed it without issue using make. Now, I get this error instead:
`
In file included from /home/blehmeh98/games/aceofspades/betterspades/BetterSpades/src/utils.c:20:
/home/blehmeh98/games/aceofspades/betterspades/BetterSpades/src/utils.h:27:48: error: unknown type name ‘size_t’
 int int_cmp(void* first_key, void* second_key, size_t key_size);
                                                ^~~~~~
/home/blehmeh98/games/aceofspades/betterspades/BetterSpades/src/utils.h:27:48: note: ‘size_t’ is defined in header ‘<stddef.h>’; did you forget to ‘#include <stddef.h>’?
/home/blehmeh98/games/aceofspades/betterspades/BetterSpades/src/utils.h:24:1:
+#include <stddef.h>
 
/home/blehmeh98/games/aceofspades/betterspades/BetterSpades/src/utils.h:27:48:
 int int_cmp(void* first_key, void* second_key, size_t key_size);
                                                ^~~~~~
/home/blehmeh98/games/aceofspades/betterspades/BetterSpades/src/utils.h:28:1: error: unknown type name ‘size_t’
 size_t int_hash(void* raw_key, size_t key_size);
 ^~~~~~
/home/blehmeh98/games/aceofspades/betterspades/BetterSpades/src/utils.h:28:1: note: ‘size_t’ is defined in header ‘<stddef.h>’; did you forget to ‘#include <stddef.h>’?
/home/blehmeh98/games/aceofspades/betterspades/BetterSpades/src/utils.h:28:32: error: unknown type name ‘size_t’
 size_t int_hash(void* raw_key, size_t key_size);
                                ^~~~~~
/home/blehmeh98/games/aceofspades/betterspades/BetterSpades/src/utils.h:28:32: note: ‘size_t’ is defined in header ‘<stddef.h>’; did you forget to ‘#include <stddef.h>’?
/home/blehmeh98/games/aceofspades/betterspades/BetterSpades/src/utils.c:56:48: error: unknown type name ‘size_t’
 int int_cmp(void* first_key, void* second_key, size_t key_size) {
                                                ^~~~~~
/home/blehmeh98/games/aceofspades/betterspades/BetterSpades/src/utils.c:56:48: note: ‘size_t’ is defined in header ‘<stddef.h>’; did you forget to ‘#include <stddef.h>’?
/home/blehmeh98/games/aceofspades/betterspades/BetterSpades/src/utils.c:21:1:
+#include <stddef.h>
 
/home/blehmeh98/games/aceofspades/betterspades/BetterSpades/src/utils.c:56:48:
 int int_cmp(void* first_key, void* second_key, size_t key_size) {
                                                ^~~~~~
/home/blehmeh98/games/aceofspades/betterspades/BetterSpades/src/utils.c:60:1: error: unknown type name ‘size_t’
 size_t int_hash(void* raw_key, size_t key_size) {
 ^~~~~~
/home/blehmeh98/games/aceofspades/betterspades/BetterSpades/src/utils.c:60:1: note: ‘size_t’ is defined in header ‘<stddef.h>’; did you forget to ‘#include <stddef.h>’?
/home/blehmeh98/games/aceofspades/betterspades/BetterSpades/src/utils.c:60:32: error: unknown type name ‘size_t’
 size_t int_hash(void* raw_key, size_t key_size) {
                                ^~~~~~
/home/blehmeh98/games/aceofspades/betterspades/BetterSpades/src/utils.c:60:32: note: ‘size_t’ is defined in header ‘<stddef.h>’; did you forget to ‘#include <stddef.h>’?
make[2]: *** [src/CMakeFiles/client.dir/build.make:427: src/CMakeFiles/client.dir/utils.c.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:91: src/CMakeFiles/client.dir/all] Error 2
make: *** [Makefile:84: all] Error 2
`

It's all about size_t now, GLFW_RAW_MOUSE_MOTION is sooooo last year.
It's also worth noting that if I make a new build directory, I can't use cmake if I don't have the libglfw package installed, because it complains it's not installed, even if I did a make install on the original source. I need to have the debian package installed in order to use cmake to get a build directory going, and I need to have the fresh source compilation installed in order to get this error instead of the GLFW_RAW_MOUSE_MOTION error.

What do I do now?

**Comment by xtreme8000** [comment]
Just follow the recommendation of your compiler, add `#include <stddef.h>` to `utils.h`.

**Comment by chandler-latour** [comment]
Did it now. New error.
`
make[2]: *** No rule to make target '/usr/lib/x86_64-linux-gnu/libglfw.so', needed by 'BetterSpades/client'.  Stop.
make[1]: *** [CMakeFiles/Makefile2:91: src/CMakeFiles/client.dir/all] Error 2
make: *** [Makefile:84: all] Error 2
`

Dunno why it's complaining that there's no libglfw.so if it uses libglfw3 and not 1. I don't really know what to do now. 
I'm sorry that crap keeps happening and I don't know how to fix it, and I really appreciate all the advice you've given me so far.

EDIT:
Oh, I forgot to mention, this is the error I get if I only have the version installed that I compiled from source. If I have the debian one installed too, it brings up a different error, another one about the raw mouse motion thing. It gets all the way to the linking stage before it complains about it too.

**Comment by xtreme8000** [comment]
You can also use GLFW from debian, you'd need to remove the two lines in `window.c` where it complains about the missing raw mouse stuff. Which is the other error you get when using it?

**Comment by chandler-latour** [comment]
`
/usr/bin/ld: CMakeFiles/client.dir/window.c.o: in function `window_init':
window.c:(.text+0x662): undefined reference to `glfwRawMouseMotionSupported'
collect2: error: ld returned 1 exit status
make[2]: *** [src/CMakeFiles/client.dir/build.make:616: BetterSpades/client] Error 1
make[1]: *** [CMakeFiles/Makefile2:91: src/CMakeFiles/client.dir/all] Error 2
make: *** [Makefile:84: all] Error 2
`

**Comment by chandler-latour** [comment]
I took your advice and I commented out lines 176 and 177, the part where it checks if the glfw instance supports raw mouse input and sets it if so. I compiled the game, and it compiled. I launched it, the menu worked great, I hopped into a server expecting mouse input to be weird (you know, where it doesn't latch to the center of the screen and disappear so that you're moving the mouse to look around and you click outside the window), but it worked flawlessly! I tried testing it with other uses, such as focusing and unfocusing the window, alt tabbing in and out of the game, all of that, but it continued to work as I would expect it to. 
Thanks!
Also, I've been reading a bit of the code, and I would like to thank you for writing this thing! I might use it to learn a little bit of glfw myself!
